messaging-spec.md

```diff
@@ -0,0 +1,351 @@
+		notify: impl Into<<T as Config>::RuntimeCall>,
+		timeout: BlockNumberFor<T>,
+		match_querier: impl Into<Location>,
+	) -> u64 {
+        // snip 
+    }
+```
+Unfortunately the QueryHandler trait doesnt contain the notify parameter hence we have to use a wrapper trait: 
+```rust
+pub trait NotifyQueryHandler<T: Config> {
+	/// Attempt to create a new query ID and register it as a query that is yet to respond, and
+	/// which will call a dispatchable when a response happens.
+	fn new_notify_query(
+		responder: impl Into<Location>,
+		notify: Call<T>,
+		timeout: BlockNumberOf<T>,
+		match_querier: impl Into<Location>,
+	) -> u64;
+}
+```
+
+This returns us a XCM query_id, not to be confused with our message_id, which we can use to match to a request and therefore the registered call back.
+When pallet-xcm recieves the response it will decode the runtime call and call our extrinsic for response handling.
+
+### ISMP and Queries
+We can use the `IsmpDispatcher` to dispatch all outgoing calls.
+
+### Message IDs
+Message ids are used as a key for outgoing requests, they can be queried to find the status of a message.
+A message id can either be specified by the runtime or the contract.
+
+If the message id is specified by the runtime, the runtime will have to return that data to the contract to emit or emit an event itself. The event emitted must contain data that can relate the request to a response.
+
+If the message id is specified by the contract, the contract developer has full control over the shape of the id and can emit a reference that can be used to link a request with a response. This must be considered by the contract developer therefore, one can create a library for ink developers (both for the contract and for the app side) that allows contract developers to easily match responses and requests.
+
+If the latter is to be implemented we should at least prevent message_id collisions in runtime storage. This will improve dev ex. 
+Hence i propose we append/prepend the contract address onto the message id. We must also use a cryptographically secure hasher due to the ids being user input.
+
+I propose we let the contract develper define a message id.
+
+TLDR: 
+- Allow developers to specify a message ID.
+- append/prepend the contract address to the message id.
+- Use a cryptographically secure hasher.
+
+### RequestDeposits
+For each outgoing request we will take a request deposit from the caller.
+This will be calculated by the request encoded len  *  byte fee. (TODO: more clarification)
+
+
+### Extrinsics
+See "Weight and Benchmarking" for a walkthrough of weight charging.
+
+```rust 
+        /// Send an ismp get message and optionally specify a callback.
+		#[pallet::call_index(1)]
+		#[pallet::weight(
+			T::WeightInfo::ismp_get() + 
+            T::IsmpModuleWeights::on_response() + 
+            callback.map(|cb| cb.weight + T::CallbackExecutor::execution_weight()).unwrap_or_default()
+		)]
+		pub fn ismp_get(
```

Comment:
is it possible to have just a few high-level dispatchables?

Something like:
```
// pseudo code
enum ProtocolMessage;
enum ProtocolResponse;

fn send(message: ProtocolMessage);
fn query(message_id) -> ProtocolResponse;
fn remove(message_id);
```

The higher level architecture might look like:
<img width="1249" alt="Screenshot 2025-02-17 at 10 31 48 AM" src="https://github.com/user-attachments/assets/ce4ed1f9-80ca-4583-acc3-7282aefabf83" />

Reply:
ill have something liek this ready for our call